app/vlinsert/journald: fix stream parsing and -maxConcurrentInserts

MatthiasGrandl opened the pull request:
### Describe Your Changes

This commit (63dccea9324037aad110cb6b7da023f0bd6a3e84) is broken. It breaks both line parsing and -maxConcurrentInserts . This PR fixes both.

Before:
```
2025-06-19T03:23:55.137Z        warn    app/vlinsert/journald/journald.go:148   remoteAddr: "127.0.0.1:62656"; requestURI: /insert/journald/upload; cannot read journald protocol data: remoteAddr="127.0.0.1:62656", requestURI="/insert/journald/upload": unexpected "_SELINUX_CONTEXT" value size; got 10 bytes; want 11 bytes; value: "unconfined"
```
with -maxConcurrentInserts=1
<img width="1800" alt="Screenshot 2025-06-19 at 05 40 31" src="https://github.com/user-attachments/assets/a2aee2c9-4d15-4278-b46f-9873789a4f7a" />

After:

<img width="1800" alt="Screenshot 2025-06-19 at 05 41 33" src="https://github.com/user-attachments/assets/d962470a-3cd8-4e53-ac73-d8a434006184" />



### Checklist

The following checks are **mandatory**:

- [ ] My change adheres to [VictoriaMetrics contributing guidelines](https://docs.victoriametrics.com/victoriametrics/contributing/#pull-request-checklist).

septatrix commented on app/vlinsert/journald/journald.go:
```diff
@@ -296,7 +295,7 @@ func readJournaldLogEntry(streamName string, lr *insertutil.LineReader, lmp inse
 			}
-			value = bytesutil.ToUnsafeString(fb.value[8 : len(fb.value)-1])
+			value = bytesutil.ToUnsafeString(fb.value[8 : 8+size])
 			if uint64(len(value)) != size {
```
With this change the check is now impossible to ever be triggered

func25 commented:
Hi, the `wcr.DecConcurrency()` prevents CPU overload, not only connection count limits. Removing this call breaks the intended concurrency control mechanism. 

Your results seem to come from a synthetic benchmark; benchmark can help expose performance bottlenecks, we must avoid optimizing for edge cases at the expense of general reliability and correctness. The real-world workloads generally involve many idle connections with low traffic, which the current design handles efficiently. Though the underlying issue you observed (high memory usage under specific high-concurrency, high-throughput benchmarks) is worth investigating further.

The second part of your PR silently trims values instead of checking their validity. This bypasses important safety checks and could mask data corruption or encoding errors, which is risky.

MatthiasGrandl commented:
> The second part of your PR silently trims values instead of checking their validity. This bypasses important safety checks and could mask data corruption or encoding errors, which is risky.

The parsing errors come from a real journal though, so master branch is definitely broken.

> Hi, the wcr.DecConcurrency() prevents CPU overload, not only connection count limits. Removing this call breaks the intended concurrency control mechanism.

> Your results seem to come from a synthetic benchmark; benchmark can help expose performance bottlenecks, we must avoid optimizing for edge cases at the expense of general reliability and correctness. The real-world workloads generally involve many idle connections with low traffic, which the current design handles efficiently. Though the underlying issue you observed (high memory usage under specific high-concurrency, high-throughput benchmarks) is worth investigating further.

I don’t understand this. The stresstest is modeled after a very real real world scenario that we encountered in production. And on the current master branch maxConcurrentInserts actively does nothing. It doesn’t matter whether you set it to 1 or 128 you always get the same level of concurrency. I don’t see how any extra CPU overload mechanism would completely break maxConcurrentInserts which sounds very explicit.

func25 commented:
@AndrewChubatiuk @valyala Please share your input on the journald changes.

MatthiasGrandl commented:
> Hm it's about how the concurrency limiter is intended to work versus what the proposed change actually does. The limiter is intentionally only activated after reading the first byte from the input stream. This allows VictoriaLogs to support **many** idle or low-throughput connections. `wcr.DecConcurrency()` is used to release the slot after processing each log line. Removing it means the slot is never released during streaming, which turns the concurrency limit into a cap on open connections, not active processing.

Ah ok that makes sense, then I guess the issue is that “processing” in this case only refers to parsing the journald log line while my stress test accumulates many of these and then gets bottlenecked at the vlstorage layer.

AndrewChubatiuk commented:
can say nothing regarding concurrency, but journal parsing really seems to be broken, checking it now

septatrix commented:
> can say nothing regarding concurrency, but journal parsing really seems to be broken, checking it now

The problem is the newline handling (adding, or rather stripping thereof) when the field value ends with one.

```txt
septatrix@Framework-13-AMD ~ [SIGINT]> /usr/lib/systemd/systemd-journal-upload --url http://localhost:9428/insert/journald --cursor=(journalctl --since=today -t python3 -n1 -o json-pretty | jq -r .__CURSOR) --follow=no
No more entries, closing journal.
Upload to http://localhost:9428/insert/journald/upload failed with code 400: remoteAddr: "127.0.0.1:55764"; requestURI: /insert/journald/upload; cannot read journald protocol data: remoteAddr="127.0.0.1:55764", requestURI="/insert/journald/upload": unexpected "BAR" value size; got 4 bytes; want 5 bytes; value: "test"
septatrix@Framework-13-AMD ~ [1]> journalctl --since=today -t python3 -n1 -o json-pretty
{
        "CODE_LINE" : "1",
        "_HOSTNAME" : "Framework-13-AMD",
        "_MACHINE_ID" : "b4735aed693e4e5d8062444deb7327c8",
        "__CURSOR" : "s=6559a11538a04164aed51c2aff3577b3;i=7f8d96;b=31c1b315b4ef4b3d8ab9e909b17697da;m=e0867acb4;t=637e996f00cb1;x=4b9c338e3e468341",
        "MESSAGE" : "test",
        "_CMDLINE" : "python3",
        "_TRANSPORT" : "journal",
        "_SYSTEMD_CGROUP" : "/user.slice/user-1000.slice/user@1000.service/app.slice/app-org.kde.konsole@acfdff1d1e3e4e7fa59c8cbed0c69ed4.service",
        "__MONOTONIC_TIMESTAMP" : "60270554292",
        "_SYSTEMD_USER_SLICE" : "app.slice",
        "__REALTIME_TIMESTAMP" : "1750326259485873",
        "SYSLOG_IDENTIFIER" : "python3",
        "BAR" : "test\n",
        "_EXE" : "/usr/bin/python3.13",
        "_AUDIT_SESSION" : "3",
        "FOO" : "a\nb",
        "_SOURCE_REALTIME_TIMESTAMP" : "1750326259485812",
        "_SYSTEMD_UNIT" : "user@1000.service",
        "_GID" : "1000",
        "_CAP_EFFECTIVE" : "0",
        "__SEQNUM_ID" : "6559a11538a04164aed51c2aff3577b3",
        "_RUNTIME_SCOPE" : "system",
        "_SYSTEMD_USER_UNIT" : "app-org.kde.konsole@acfdff1d1e3e4e7fa59c8cbed0c69ed4.service",
        "_SYSTEMD_OWNER_UID" : "1000",
        "CODE_FUNC" : "<module>",
        "BINARY" : [
                0,
                98,
                108,
                111,
                98,
                195,
                157,
                0
        ],
        "_UID" : "1000",
        "_SYSTEMD_SLICE" : "user-1000.slice",
        "_AUDIT_LOGINUID" : "1000",
        "_BOOT_ID" : "31c1b315b4ef4b3d8ab9e909b17697da",
        "__SEQNUM" : "8359318",
        "_PID" : "221196",
        "CODE_FILE" : "<python-input-5>",
        "_COMM" : "python3",
        "_SYSTEMD_INVOCATION_ID" : "7ff6a9f5cad746ac9c93c67bc920dc4a",
        "_SELINUX_CONTEXT" : "unconfined_u:unconfined_r:unconfined_t:s0-s0:c0.c1023"
}
```

generated using python

```python
import systemd.journal
systemd.journal.send("test", FOO="a\nb", BAR="test\n", BINARY="\0blob\xDD\0")
```

The loop for receiving the length field is also either broken or completely redundant (`for len(fb.value) < 8 {`). If there ever were a case where it is executed twice then there is a newline mangled in there which would lead to a wrong result of `binary.LittleEndian.Uint64`

AndrewChubatiuk commented:
added fix for message parsing [here](https://github.com/VictoriaMetrics/VictoriaMetrics/pull/9234/files)

MatthiasGrandl commented:
> @MatthiasGrandl That's strange, so you tested with `-maxConcurrentInserts=1`, removed the decrement, and it worked fine with 86 concurrent streams?
> 
> Edit: Ok, I misunderstood. You tried -maxConcurrentInserts=1 and it worked fine, that's expected. Could you please remove the decrement (as your PR) and try again?

No you understood right. I removed the dec and it worked with 86 concurrent streams.

valyala commented:
> we have seen it in production and it is easy to trigger. All it takes is a 50+ nodes writing to victorialogs and then the victorialogs node going down for maintenance or something. When it comes back it will be hammered exactly like this, causing a loop that is hard to recover from

@MatthiasGrandl , the graphs provided by you above suggest that you run VictoriaLogs on a host with 36GiB of RAM. VictoriaLogs uses 3.5% of RAM, e.g. `36GiB*3.5%=1.26GiB` with `-maxConcurrentInserts=1`. When you remove the `wcr.DecConcurrency()` call from the loop which reads Journald logs one-by-one, VictoriaLogs lowers its' memory usage to 1.4% of RAM, e.g. `36GiB*1.4%=504MiB`. These numbers of RAM usage are expected for the stress test, which hammers VictoriaLogs  with big number of logs sent over big number of concurrently established connections. These numbers are OK for the production, since the memory usage shouldn't grow too much beyond the measured RAM usage numbers for the given `-maxConcurrentInserts=1` value, even if the logs will be sent over 1000 concurrent connections instead of 50 concurrent connections.

I believe there is a room for reducing RAM usage in this case, and we should try reducing RAM usage for this case (since everybody will benefit from the reduced RAM usage). The proper approach is to collect memory profiles from VictoriaLogs instance under this stress test according to [these docs](https://docs.victoriametrics.com/victorialogs/#profiling), and then try to reduce the memory usage according to these profiles. @MatthiasGrandl , could you collect memory profiles and attach them here, so we could investigate them and try reducing RAM usage according to these profiles?

Let me explain why the removal of `wcr.DecConcurrency()` call from the loop, which reads Journald logs one-by-one, is a bad idea for production. When the `wcr.DecConcurrency()` call is removed, then this efficiently limits the number of concurrently processed connections to the `-maxConcurrentInserts`. The rest of connections just wait their turn until some of the concurrently processed connection is closed. E.g. when you set `-maxConcurrentInserts=1`, then VictoriaLogs reads logs only from a single connection until it is closed, and only after that it switches to reading the logs from another connection out of the remaining 49 established connections, and so on. Measure the number of sent logs per each concurrently established connection in your stress test in order to see that only a single connection is processed, while the rest of connection are waiting for their turn (which may never occur if the processed connection is never closed). This is not what most users expect from VictoriaLogs server when they establish big number of connections to it and send logs over them whenever these logs are emitted by some apps. Most users expect that these logs will be consumed by VictoriaLogs as soon as possible. These expectations aren't met in the case with the removed `wcr.DecConcurrency()` call, since logs will be read by VictoriaLogs from only `-maxConcurrentInserts` concurrent connections, while the rest of concurrent connections will sit idle until one of the active connection is closed (which may never happen in real life when clients establish long-running connections and send logs in small chunks over them).

MatthiasGrandl commented:
Even in non-stresstest scenarios it seems that excessive multithreading is causing unnecessary resource usage. I understand that victorialogs users expect logs to be ingested as fast as possible, but as I said in the other PR we still ingested hundreds of thousand log lines single per second with the PR flag set to 1 and when deployed into production with lots of mostly idle streams we saw quite a reduction of both CPU and RAM. I can do more profiling, but for now I am on vacation.

i just think it’s a shame that victorialogs as it stands is much more resource hungry than journald-remoted for the same workload.

MatthiasGrandl commented:
> > But the approach with the removal of the `wcr.DecConcurrency()` proposed in this pull request isn't the correct solution to this problem.
> 
> I think what Matthias was trying to get your opinion on is the approach he brought forth in #9190

@valyala correct. Sorry for getting this all confused. Let’s close this PR, I totally agree that this approach is wrong.

valyala commented:
@MatthiasGrandl , please create a separate issue about high memory usage of VictoriaLogs when it accepts Journald logs from many concurrent connections at a high rate, and attach memory profiles to the issue, so they could be investigated further.
